src/sentry/taskworker/scheduler.py

```diff
@@ -0,0 +1,255 @@
+        scheduler: Schedule
+        if isinstance(schedule, timedelta):
+            scheduler = TimedeltaSchedule(schedule)
+        else:
+            scheduler = CrontabSchedule(schedule)
+        self._schedule = scheduler
+        self._last_run: datetime | None = None
+
+    def __lt__(self, other: ScheduleEntry) -> bool:
+        # Secondary sorting for heapq when remaining time is the same
+        return self.fullname < other.fullname
+
+    @property
+    def fullname(self) -> str:
+        return self._task.fullname
+
+    def set_last_run(self, last_run: datetime | None) -> None:
+        self._last_run = last_run
+
+    def is_due(self) -> bool:
+        return self._schedule.is_due(self._last_run)
+
+    def remaining_delta(self) -> timedelta:
+        return self._schedule.remaining_delta(self._last_run)
+
+    def runtime_after(self, start: datetime) -> datetime:
+        return self._schedule.runtime_after(start)
+
+    def delay_task(self) -> None:
+        self._task.delay()
+
+
+class ScheduleSet:
+    """A collection of ScheduleEntry objects"""
+
+    def __init__(self, registry: TaskRegistry, run_storage: RunStorage) -> None:
+        self._entries: list[ScheduleEntry] = []
+        self._registry = registry
+        self._run_storage = run_storage
+        self._heap: list[tuple[float, ScheduleEntry]] = []
+
+    def add(self, task_config: ScheduleConfig) -> None:
+        """Add a task to the schedule."""
+        try:
+            (namespace, taskname) = task_config["task"].split(":")
+        except ValueError:
+            raise ValueError("Invalid task name. Must be in the format namespace:taskname")
+
+        task = self._registry.get_task(namespace, taskname)
+        entry = ScheduleEntry(task=task, schedule=task_config["schedule"])
+        self._entries.append(entry)
+        self._heap = []
+
+    def tick(self) -> float:
+        """
+        Check if any tasks are due to run at current_time, and spawn them.
+
+        Returns the number of seconds to sleep until the next task is due.
+        """
+        self._build_heap()
+
+        while True:
+            # Peek at the top, and if it is due, pop, spawn and update.
+            _, entry = self._heap[0]
+            if entry.is_due():
+                heapq.heappop(self._heap)
+                now = timezone.now()
+                next_runtime = entry.runtime_after(now)
+                if self._run_storage.set(entry.fullname, next_runtime):
+                    entry.delay_task()
+                entry.set_last_run(now)
+                heapq.heappush(self._heap, (entry.remaining_delta().total_seconds(), entry))
+                continue
+            else:
+                # The top of the heap isn't ready, break for sleep
+                break
+
+        return entry.remaining_delta().total_seconds()
+
+    def _build_heap(self) -> None:
+        if self._heap:
+            return
+
+        heap_items = []
+        for item in self._entries:
```

Comment:
Does it make sense to refactor this to use redis pipelines? Instead of making individual calls?

Reply:
Good idea! I've done that now.